### PR TITLE
Virology mathmatical rebalancing.

### PR DESCRIPTION
**Pull request by LanCartwright**
:cl: 
tweak: New formula for toxic filter healing. New formula is SQRT(20+X)*(2+N) Where X equals the total stage speed of the custom virus, and N equals a positive number less than equal to 1.
tweak: Now requires 2u of each chemical with virus food to make 1u of the resulting virus chemical.
tweak: 1u Plasma food now requires 2u Plasma and 1u of Virus food.
tweak: 1u Mutagenic agar now requires 2u Mutagen and 1u of Virus food.
tweak: Weakened virus plasma now results in 1u instead of 2u.
tweak: Sucrose agar now results in 1u instead of 2u.
/:cl:

### CODE REVIEW

**Comment by Buggy123** [comment]
All it took was a few people who were good at it, and it's nerfed.

**Comment by LanCartwright** [comment]
There's a difference between good at it, and to the point where it disrupts the round.
i.e. giving out hulk to assistants every round for a week resulted in the hulk nerf.

Likewise away missions are now gone not because "a few people got good", but rather they chose to do it until the hand was forced.

**Comment by PKPenguin321** [comment]
:-1: makes virology even more tedious, toxic filter may be warranted though

**Comment by LanCartwright** [comment]
Not really.
It's negligible, considering previous to this the odds for a level 6 symptom was 1/22, currently it's about 1/11 with the new chems.

**Comment by Robustin** [comment]
Only nerf needed was only allowing for one virus at a time. Give viruses a stat that would decide if they overtake and replace weaker viruses. 

Healing virus is ok. 3x healing virus is close to fleshmend levels of healing.

**Comment by LanCartwright** [comment]
There is no damage per second healing. You won't find it.

**Comment by kevinz000** [comment]
What's the point of buffing it by making it not RNG and therefore easy to make super powerful viruses and then nerfing the shit out of it? Revert your other PR and we're golden as most people took 20-30 minutes before that.

**Comment by LanCartwright** [comment]
No. They didn't. If they did, it was unlucky RNG or slowness.

Timings between the update and before the update was within a 3 minute difference; slower than the cargo shuttle.

previous update did not change how the viruses worked at all. The toxic filter nerf is long overdue.

I'd suggest looking at how this works carefully, since this is not clear cut a nerf. At the board of stats, it's possible to still make a healing virus stronger than before.

**Comment by Core0verload** [comment]
:-1: stop nerfing the only thing that stops virologist from releasing horrible death on everyone

File: code/datums/diseases/advance/advance.dm [hunk 1]
```diff
@@ -424,4 +424,9 @@ var/list/advance_cures = 	list(
 		src << "<a href='?_src_=vars;Vars=\ref[D]'>[D.name] - [D.holder]</a>"
 */
 
+/datum/symptom/heal/proc/Heal(mob/living/M, datum/disease/advance/A)
+	var/get_damage = sqrt(20-"stage_rate")*(2+rand())
```
Comment (KazeEspada) [inline code/datums/diseases/advance/advance.dm]: "stage_rate" is a string? unless you are doing black magic of some sort.

**Comment by AlexCrimson** [comment]
No idea what half this shit is. Why do people not update the wiki when they radically change a department?

**Comment by LanCartwright** [comment]
@alexcrimson that's my job to do, and I'll be updating it today.

File: code/datums/diseases/advance/advance.dm (hunk 1 above)
Comment (LanCartwright) [inline code/datums/diseases/advance/advance.dm]: It uses the stage_rate code from above (hopefully) since I've had some difficulties finding the code for the total stage speed of viruses which has been an annoyance.

Comment (Giacom) [inline code/datums/diseases/advance/advance.dm]: I'm guessing "stage_rate" is a key for the ```properties``` associative list. You're not accessing that list though, you're just getting the square root of

```20 - "stage_rate"```

Which makes no sense since you're doing a minus operation to a string.

**Comment by newfren** [comment]
:-1: I don't want to do complex math every time I make a healing virus to see if it's actually any good.

**Comment by CPTANT** [comment]
:-1: 
Makes virology completely irrelevant.

**Comment by LanCartwright** [comment]
Added new virology chems to the wiki.

I never wanted to do this, but my hand was forced.

I'm going to do some simple comparisons for you;
If this nerf goes through, on average (assuming perfect random) the following virus setups will be:
Self respiration, toxic filter, toxic compensation, stimulants, eternal youth.
Total stage speed = -9
75.38% effective.
24.62% nerf.

Toxic filter, Toxic compensation, eternal youth
Total stage speed = -4
90.91% effective.
9.09% nerf.

File: code/datums/diseases/advance/advance.dm (hunk 1 above)
Comment (LanCartwright) [inline code/datums/diseases/advance/advance.dm]: I don't quite know to to test it @Giacom, my apologies. I'll try and setup a test virus using this code.

I'm trying to figure out how to find the total stage_speed to no luck.

Do you have any suggestions to getting the total stage_speed of a custom virus?

Comment (Giacom) [inline code/datums/diseases/advance/advance.dm]: You can loop through the viruses' symptoms and add the stage_speed of each one.

Comment (Giacom) [inline code/datums/diseases/advance/advance.dm]: I would also recommend reading about BYOND's associative lists. Which, in other languages, are called a HashMap or Dictionary.

Difference between these two kinds of lists is that, with the normal list, you can access the element by the order it is in the list:

```C++
var/list/my_list = list()
my_list += 5
my_list[1] // Would return 5, since it is the 1st in the list
```

And with the associative list, which is what the ```properties``` variable is, you can add your values differently and retrieve them in any order. You just need to give it a key so it knows what value to give you. Like so:

```C++
var/list/my_list = list()
my_list["my_key"] = 5
my_list["my_key"] // Will give you back your 5.
```

There is a shorter way to write this, like so:

```C++
var/list/my_list = list("my_key" = 5)
```

Which is what I do in the advance.dm file.

**Comment by Robustin** [comment]
I know the viruses don't come with a healing/time effect.

I'm just saying you can TEST it and give feedback.

How much healing does 3x stacking the best healing virus CURRENTLY provide?

How much healing does 3x stacking the best virus provide under this PR?

**Comment by LanCartwright** [comment]
Here's the current:
Currently the best healing viruses provide:
N*(S*10)
Where N is random between 8 and 14.
And S is SYMPTOM_ACTIVATION_PROB.

Under this system, the healing is:
T*(S*10)
Where T is the mean square root of the total stage speed of all symptoms of the virus plus 20, multiplied by a random number between 2 and 3.

Under the new system, a virus with the LOWEST theoretical stage speed, and heals would be:
Hyphema -4
Spontaneous combustion -4
Toxic Compensation -4
Toxic filter -4
Confusion -3
-19

The highest theoretical speed and with healing symptoms would be:
Viral Evolutionary acceleration +5
Eternal youth +5
Longevity +5
Toxic compensation -4
Toxic filter -4
+7 

2.5(S*10) will be the average lowest possible healing rate.
13(S*10) will be the average highest possible healing rate.

In comparison to previous nerf
11(S*10) was the average healing rate, flat across the board. Nothing affected this healing rate.

**Comment by Buggy123** [comment]
I don't disagree with you but I see one fatal flaw here. Lets actually calculate some values using the equation copy-pasted from the code. I'll assume random is just 1 for simplicity.

sqrt(20-"stage_rate")*(2+rand())

stage speed -8 (toxic comp+filt)
sqrt(20-(-8))*(2+(1))=sqrt(20+8)*(3)=Sqrt(28)*3~15.9


stage speed +7
sqrt(20-(7))*(2+(1))=sqrt(17)*(3)=12.4~

See the problem here?

**Comment by LanCartwright** [comment]
oh shit. Nice catch on that. It should be 20+stage_rate.
+7 = 
20+7 = 27
-7=
20+-7 = 13

**Comment by RemieRichards** [comment]
DNM by request.

**Comment by AlexCrimson** [comment]
I do not see the big deal. Healing viruses have always been insanely strong, especially when you release 3 of them. My issue with Virology is that healing viruses are about the only useful thing you can do as a non-traitor, but thats an issue for another PR. 

The healing nerf was needed. The wiki was updated. Im happy.

**Comment by Jazaen** [comment]
At least don't nerf viro chems, since you are already healing toxic comp.

**Comment by LanCartwright** [comment]
image

Here is a proposed buff to Necrotizing Fasciitis using an alternative calculation of sqrt(20+A.totalStageSpeed())*(1+rand())

File: code/datums/diseases/advance/advance.dm [hunk 2]
```diff
@@ -424,4 +424,16 @@ var/list/advance_cures = 	list(
...
+	var/total_stage_speed = 0
+	for(var/i in symptoms)
+		var/datum/symptom/S = i
+		total_stage_speed += S.stage_speed
+	return total_stage_speed
+
+/datum/symptom/heal/proc/Heal(mob/living/M, datum/disease/advance/A)
+	var/get_damage = sqrt(20+A.totalStageSpeed())*(2+rand())
+	M.adjustToxLoss(-get_damage)
+	return 1
```
Comment (phil235) [inline code/datums/diseases/advance/advance.dm]: why did you move the Heal proc here?

Comment (LanCartwright) [inline code/datums/diseases/advance/advance.dm]: something to do with errors and undefined vars i think.

Comment (phil235) [inline code/datums/diseases/advance/advance.dm]: I don't see how moving a proc has anything to do with undefined vars and error, when you are not using any defines.

Comment (LanCartwright) [inline code/datums/diseases/advance/advance.dm]: I think I did this before the new code was implemented. I'll move it back if necessary or if it's better that way.

**Comment by LanCartwright** [comment]
DNM can be removed since Giacom's mentioned issues are fixed.

**Comment by LanCartwright** [comment]
You can visit here to change variables and see the changes:
https://docs.google.com/spreadsheets/d/11yPq6BSPvCKzwusxt7YrUhM9_0oY0ZT-Q2cdgt2Wr6g/edit?usp=sharing

File: code/datums/diseases/advance/symptoms/fire.dm [hunk 3]
```diff
@@ -33,15 +33,25 @@ Bonus
...
+				Firestacks2(M, A)
 				M.IgniteMob()
 				M << "<span class='userdanger'>Your skin erupts into an inferno!</span>"
-				M.adjustFireLoss(10)
 				M.emote("scream")
-	return
\ No newline at end of file
+	return
+
+/datum/symptom/fire/proc/Firestacks1(mob/living/M, datum/disease/advance/A)
```
Comment (Giacom) [inline code/datums/diseases/advance/symptoms/fire.dm]: It is a bad idea to have function names like this. I have no idea what Firestacks1 does differently compared to Firestacks2, without having to read the code myself. You should consider a more descriptive name or..

You can merge these functions together, since they contain duplicate code. You can read CONTRIBUTING.md about why duplicated code like this is bad.

Comment (LanCartwright) [inline code/datums/diseases/advance/symptoms/fire.dm]: Ah, so I should have then with better descriptive names i.e. stage4firestack and stage5firestack or make two functions that use get_stacks_stage_4 and get_stacks_stage_5 instead?

Comment (Cheridan) [inline code/datums/diseases/advance/symptoms/fire.dm]: Just make it 1 proc that accepts an argument
